lib/matplotlib/figure.py

```diff
@@ -1522,6 +1523,197 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
                 .subplots(sharex=sharex, sharey=sharey, squeeze=squeeze,
                           subplot_kw=subplot_kw))
 
+    @staticmethod
+    def _normalize_grid_string(layout):
+        layout = inspect.cleandoc(layout)
+        return [list(ln) for ln in layout.strip('\n').split('\n')]
+
+    def subplot_mosaic(self, layout, *, subplot_kw=None, gridspec_kw=None,
+                       empty_sentinel='.'):
+        """
+        Build a layout of Axes based on 2D layout.
+
+        This is a helper function to build complex GridSpec layouts visually.
+
+        .. note ::
+
+           This API is provisional and may be revised in the future based on
+           early user feedback.
+
+
+        Parameters
+        ----------
+        layout : list of list of {hashable or nested} or str
+
+            A visual layout of how you want your Axes to be arranged
+            labeled as strings.  For example ::
+
+               x = [['A', 'A', 'edge'],
+                    ['C', 'D', 'edge']]
+
+            Produces 4 axes:
+
+            - 'A' which is 1 row high and spans the first two columns
+            - 'edge' which is 2 rows high and is on the right edge
+            - 'C' which in 1 row and 1 column wide in the bottom left
+            - 'D' which in 1 row and 1 column wide in the bottom center
+
+            If input is a str, then it must be of the form ::
+
+              '''
+              AAB
+              C.D
+              '''
+
+            where the labels are assumed to be single characters and
+            the rows are delimited by new lines.  This only allows
+            single character axes labels and does not allow nesting,
+            but is very terse.
+
+            Using nested lists any hashable objects can be used to label
+            the Axes.  In the special case that the objects in the list are
+            existing Axes objects they adjusted to match the requested layout.
```

Comment:
"adjusted" -> "are adjusted"

Reply:
Not sure I like passing existing Axes to rearrange them. This feels a bit like trying too much in the function.

While it looks neat at first, it adds complexity:

- To the code: It's now just a few lines, but are we sure this is working reasonably? E.g. what happens for `layout=[[ax, '.', ax]]`?
- To the semantics: "create or rearrange" instead of just "create". Technically that would have to be reflected in various places, e.g. again the return value

  > A dictionary mapping the string labels to the new Axes objects.

  is twice imprecise: it's neither a string label nor a new Axes.

- The returned dict mapping `ax -> ax` is a bit odd.

Coming from the other side: Is there a particular user need to rearrange complex axes? I'd think this is very rare.

Following KISS, I would not support rearranging in the first experimental version. We can always add it later.